### PR TITLE
Streamflow Diags Enhancements

### PR DESCRIPTION
**Pull request by forsyth2**
Streamflow Diags enhancements

- Specifications for data paths for test and ref `.nc` and `.mat` files.
- Improve plot titles.

### CODE REVIEW

File: demo_run.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# TODO: Remove before merging
```
Comment (forsyth2) [inline demo_run.py:1]: Remove this file before merging.

**Comment by forsyth2** [comment]
@chengzhuzhang With these changes, the user only has to specify the test and ref data paths rather than a specific `.mat` file or `.csv` file. There are a few caveats:
- The gauge file (the `.csv` file) must be in a subdirectory of `parameter.reference_data_path` called `gauges`
- If the user wants to use a `.mat` file rather than a `.nc` file, then `parameter.reference_data_path` must end with `mat`. Furthermore, there can only be one `.mat` file in that directory.

My data is currently stored locally as: 
```
-3yr_nc/
--RIVER_DISCHARGE_OVER_LAND_LIQ_195901_196112.nc
--gauges/
---GSIM_catchment_characteristics_all_1km2.csv

-500yr_nc/
--RIVER_DISCHARGE_OVER_LAND_LIQ_000101_050012.nc

-edison_mat/
--edison.20181204.noCNT.A_WCYCL1950S_CMIP6_LRtunedHR.ne30_oECv3_ICG_0001-0050.mat

-obs_mat/
--GSIM.mat
--gauges/
---GSIM_catchment_characteristics_all_1km2.csv

-test_mat/
--test.mat

```

If all of the above is acceptable, I can add the `GSIM.mat` file and the `GSIM_catchment_characteristics_all_1km2.csv` file to a new directory in `/global/project/projectdirs/e3sm/acme_diags/obs_for_e3sm_diags/` as follows:
```
-/global/project/projectdirs/e3sm/acme_diags/obs_for_e3sm_diags/gsim_mat/
--GSIM.mat
--gauges/
---GSIM_catchment_characteristics_all_1km2.csv
```

What are your thoughts? What other machines need the data transferred to them? I'm assuming acme1 and compy. Any others?

File: acme_diags/driver/streamflow_driver.py [hunk 2]
```diff
@@ -72,9 +73,21 @@ def get_seasonality(monthly):
...
+    if parameter.test_data_path.strip('/').endswith('mat'):
+        using_test_mat_file = True
+    using_ref_mat_file = False
+    if parameter.reference_data_path.strip('/').endswith('mat'):
+        using_ref_mat_file = True
+    # The metadata file of GSIM that has observed gauge lat lon and drainage area
+    # This file includes 25765 gauges, which is a subset of the entire
+    # dataset (30959 gauges). The removed gauges are associated with very
+    # small drainage area (<1km2), which is not meaningful to be included.
+    gauges_path = '/{}/gauges/GSIM_catchment_characteristics_all_1km2.csv'.format(parameter.reference_data_path.strip('/'))
```
Comment (forsyth2) [inline acme_diags/driver/streamflow_driver.py:86]: Note: `csv` file name is hard-coded.

File: acme_diags/driver/streamflow_driver.py [hunk 3]
```diff
@@ -99,15 +112,16 @@ def run_diag(parameter):
             # the data has been reorganized to a 1380 * 30961 matrix. 1380 is the month
             # number from 1901.1 to 2015.12. 30961 include two columns for year and month plus
             # streamflow at 30959 gauge locations reported by GSIM
-            ref_mat = scipy.io.loadmat(parameter.ref_mat_file)
+            ref_mat_file = '/{}/GSIM.mat'.format(parameter.reference_data_path.strip('/'))
```
Comment (forsyth2) [inline acme_diags/driver/streamflow_driver.py:115]: Note: reference `mat` file name is hard-coded.

File: acme_diags/driver/streamflow_driver.py [hunk 4]
```diff
@@ -123,7 +137,17 @@ def run_diag(parameter):
             if parameter.print_statements:
                 print('area_upstream dimensions={}'.format(area_upstream.shape))
         else:
-            data_mat = scipy.io.loadmat(parameter.test_mat_file)
+            files_in_test_data_path = os.listdir(parameter.test_data_path)
+            mat_files = list(filter(lambda file_name: file_name.endswith('.mat'), files_in_test_data_path))
```
Comment (forsyth2) [inline acme_diags/driver/streamflow_driver.py:141]: Note: if using `mat` files for `test`, then find all the `mat` files in this directory. There should only be one. We will use that one.

**Comment by chengzhuzhang** [comment]
@forsyth2 I would suggest to have a sub-directory `/GSIM` under `/obs_for_e3sm_diags/time-series/`: 
`/obs_for_e3sm_diags/time-series/GSIM/gauges/GSIM_catchment_characteristics_all_1km2.csv`
and 
`/obs_for_e3sm_diags/time-series/GSIM/GSIM_198601_199512.mat` (renamed GSIM.mat file)

We should expect test dataset to be `.nc` files all the time. and reference data can be `.mat` in the case of `model vs obs`, and reference to be `.nc` for `model vs model`. 

In `model vs model` case, the user should define the path to the gauges csv. Let me know if this makes sense..

**Review by forsyth2 — COMMENTED** [review]
@chengzhuzhang How does the second commit look? If those changes look good, I'll merge. Thanks!

For some reason, a `csv` file in the same directory as a `nc` file would cause issues when reading that `nc` file. If we're just keeping the `csv` file with the `mat` file or in a user-specified location, there's no need to keep it in a `gauges` subdirectory. (The code now assumes the `csv` file will be in `GSIM`. If you're not using `GSIM`, it requests the `csv` path.)

I kept the code to process `mat` files for `test`, since it's useful for debugging.

My local data is now stored as:
```
-3yr_nc/
--RIVER_DISCHARGE_OVER_LAND_LIQ_195901_196112.nc

-500yr_nc/
--RIVER_DISCHARGE_OVER_LAND_LIQ_000101_050012.nc

-edison_mat/
--edison.20181204.noCNT.A_WCYCL1950S_CMIP6_LRtunedHR.ne30_oECv3_ICG_0001-0050.mat

-GSIM/
--GSIM.mat
--GSIM_catchment_characteristics_all_1km2.csv

-test_mat/
--test.mat
```

File: acme_diags/driver/streamflow_driver.py [hunk 5]
```diff
@@ -77,13 +77,18 @@ def run_diag(parameter):
...
-    # small drainage area (<1km2), which is not meaningful to be included.
-    gauges_path = '/{}/gauges/GSIM_catchment_characteristics_all_1km2.csv'.format(parameter.reference_data_path.strip('/'))
+        # The metadata file of GSIM that has observed gauge lat lon and drainage area
+        # This file includes 25765 gauges, which is a subset of the entire
+        # dataset (30959 gauges). The removed gauges are associated with very
+        # small drainage area (<1km2), which is not meaningful to be included.
+        gauges_path = '/{}/GSIM_catchment_characteristics_all_1km2.csv'.format(parameter.reference_data_path.strip('/'))
+    else:
+        if parameter.gauges_path is None:
+            raise Exception('To use a non-GSIM reference, please specify parameter.gauges_path.')
```
Comment (forsyth2) [inline acme_diags/driver/streamflow_driver.py:89]: Perhaps further specify `streamflow_parameter.gauges_path`.

**Comment by chengzhuzhang** [comment]
I can't find my last comments. But

@forsyth2 I think the blocks reads in data (L76 -190 in `streamflow_driver.py`) may need to be straighten a bit to follow something as below:
```
#not an actual code block
if parameter.run_type = 'model-vs-model':
    Ask for gauge locations .csv file, hint on the possible location where e3sm obs data located () : ex parameter.guages = '/global/project/projectdirs/e3sm/acme_diags/obs_for_e3sm_diags/time-series/GSIM/gauges/GSIM_catchment_characteristics_all_1km2.csv' for Cori 

elif parameter.run_type = 'model-vs-obs':
     Should automatically gets the .csv file and .mat files based on ref_data_path ('/global/project/projectdirs/e3sm/acme_diags/obs_for_e3sm_diags/time-series/` take Cori for example)

else: 
      run_type not supported.
```

Let me know if this makes sense, i can try clarify.

**Comment by forsyth2** [comment]
@chengzhuzhang Thanks for the feedback. Does the third commit look good? I tested it out with `model_vs_model`, `mdoel_vs_obs`, an unsupported run type and `model_vs_model` with unspecified gauge path. 

I will work on updating the tests accordingly in #306 (updates tests to not need `nc` files in the repo)

**Comment by chengzhuzhang** [comment]
@forsyth2 Thanks for the update. 
1. for timeseries file, usually user specifies: reference_data_path = 'prefix/obs_for_e3sm_diags/time-series/'. But when i tested, i have to use `prefix/obs_for_e3sm_diags/time-series/GSIM`. 
2. Would you rename `GSIM.mat` to `GSIM_198601_199512.mat`, so that the obs file name is more consistent than others
3. The `RIVER_DISCHARGE_OVER_LAND_LIQ_195901_196112.nc` you have apparently is not a per-variable file, it includes all the variables from mosart. I contacted Charlie for using ncclimo to add areatotal2 to per-variable file.

**Comment by forsyth2** [comment]
@chengzhuzhang I added two more commits, addressing your first two comments. I specify `ref_name` as a parameter now. I tested `model_vs_model`, `model_vs_model` with no gauge path specified, and `model_vs_obs` -- all looks correct to me.

File: acme_diags/driver/streamflow_driver.py [hunk 6]
```diff
@@ -99,17 +124,20 @@ def run_diag(parameter):
             # the data has been reorganized to a 1380 * 30961 matrix. 1380 is the month
             # number from 1901.1 to 2015.12. 30961 include two columns for year and month plus
             # streamflow at 30959 gauge locations reported by GSIM
-            ref_mat = scipy.io.loadmat(parameter.ref_mat_file)
+            ref_mat_file = '/{}/GSIM/GSIM_198601_199512.mat'.format(parameter.reference_data_path.strip('/'))
+            parameter.ref_name_yrs = 'GSIM ({}-{})'.format(parameter.ref_start_yr, parameter.ref_end_yr)
```
Comment (forsyth2) [inline acme_diags/driver/streamflow_driver.py:128]: The ref name is set to be `GSIM`. Should I change this to be superseded by `parameter.short_ref_name`?

**Review by chengzhuzhang — COMMENTED** [review]
For the mean bias plot, could you use ‘Relative Bias’ as title?  After fixing this please go ahead clean up and merge. 
short_ref_name is not often used, I think your implementation is fine.

**Pull request merged**